torchtune/datasets/_sft.py

```diff
@@ -110,36 +111,61 @@ def __init__(
         if filter_fn is not None:
             self._data = self._data.filter(filter_fn)
 
+        self._prepare_sample = SFTTransform(
+            message_transform=self._message_transform,
+            model_transform=self._model_transform,
+        )
+
     def __len__(self):
         return len(self._data)
 
     def __getitem__(self, index: int) -> Dict[str, Any]:
         sample = self._data[index]
         return self._prepare_sample(sample)
 
-    def _prepare_sample(self, sample: Mapping[str, Any]) -> Dict[str, Any]:
-        transformed_sample = self._message_transform(sample)
-        if "messages" in transformed_sample:
-            validate_messages(transformed_sample["messages"])
-
-        tokenized_dict = self._model_transform(transformed_sample)
 
-        if not ("tokens" in tokenized_dict and "mask" in tokenized_dict):
-            keys_str = ", ".join(tokenized_dict.keys())
-            error_message = (
-                "model_transform returned the following keys: "
-                f"{keys_str}. Must return 'tokens' and 'mask' as keys."
+class SFTTransform(Transform):
```

Comment:
this may be the cleanest way to separate the old way and the torchdata way, but what are your thoughts on keeping the class as SFTDataset, but remove all the HF load dataset logic out? we can keep source as a class attribute but not load it, then in the recipe or a utility used in the recipe: if we are using torchdata, use the load_hf_dataset you added here. else, use a barebones load utility with the original behavior.

When we eventually migrate to torchdata, we can just remove this switch logic and all the API names will remain the same but everything is swapped out under the hood

Reply:
I think this is possible, but the tricky part is that right now, SFTDataset is a subclass of torch.utils.data.Dataset and the new version is a subclass of torchdata.nodes.BaseNode, and reconciling them may be difficult. We wouldn't be able to just make a switch because the torch.utils.DataLoader needs to be swapped out as well for the new `get_loader` builder wherever that is being called

Reply:
cc @pbontrager for thoughts